### PR TITLE
Sockets - Chantal & Tatiana

### PR DESCRIPTION
**Pull request by tatsqui**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | Method signature, parameters(if any) block, return value.
What are the advantages of using git when collaboratively working on one code base? | Easily view and edit file, without the other person (not having to email copies to each other). It tracks all changes and keeps a log of them (when comitted).
What kind of relationship did you and your pair have with the unit tests? | A tumultous one. We learned how to read unit testing a lot better and actually understand how it works and apply that to what kind of code we need to write. Helpful for debugging as well. The relationship with the tests did get better.
Does your code use any methods from the `Enumerable` mixed in? If so, where and why was it helpful? | we used a map method to return the word/score for finding out what word won.
What was one method you and your pair used to debug code? |  Pry, printing to the screen, unit tests were helpful.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We discussed out learning styles and execution styles with our coding techniques and recognized it would be great to go over that from the start and come up with a plan of action to use each other's strengths before tackling individual problems in the project.

### CODE REVIEW

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        "X",
+        "Y", "Y",
+        "Z"
+    ]
+    drawn_letters = letters.sample(10)
+    return drawn_letters
+end
+
+
+draw_letters
```
Comment (droberts-sea) [inline lib/adagrams.rb:38]: You don't need to call `draw_letters` here. Your job is to define the method, the tests and the `wave_X_game.rb` will handle calling it.

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+end
+
+
+def highest_score_from (words)
+
+   score_list = words.map do |element| 
+    { word: element, score: score_word(element) }
+  end 
+
+find_best_word = []
```
Comment (droberts-sea) [inline lib/adagrams.rb:89]: Watch your indentation through here. Have you installed rubocop and rufo to automatically adjust your formatting?

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,115 @@
+require 'awesome_print'
+require 'pry'
+
+def draw_letters
+    letters = [
+        "A", "A", "A", "A", "A", "A", "A", "A", "A", 
+        "B", "B",
+        "C", "C",
```
Comment (droberts-sea) [inline lib/adagrams.rb:8]: This data would be a great thing to store in a constant!

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+draw_letters
+
+def uses_available_letters?(input, letters_in_hand)
+    input.each_char do |letter|
+        index = letters_in_hand.index(letter)
+        if index.nil?
+            return false
+        else
+            letters_in_hand.delete_at(index)
```
Comment (droberts-sea) [inline lib/adagrams.rb:46]: When you call `delete_at` here, you modify the original hand of letters! While that makes the tests pass, it also introduces a bug:

```ruby
hand = ['A', 'B', 'C']
word = 'ABC'
puts uses_available_letters?(word, hand)
# => true
puts uses_available_letters?(word, hand)
# => false
```

How might you address this problem?

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+end
+
+def score_word (word)
+  score = 0
+  word.upcase.each_char do |letter|
+  
+  case letter
+  when "A", 'E', 'I', 'O', 'U', 'L', 'R', 'N', 'S', 'T' 
+    score += 1
+  when 'D', 'G'
```
Comment (droberts-sea) [inline lib/adagrams.rb:59]: While this `case` statement works, it means that the information about which letter has which score is locked into this piece of code, and can't easily be used elsewhere. For example, if you wanted to display the value of each letter in a hand, you would need to repeat this work.

An alternative approach would be to store the letter scores in a hash, something like this:

```ruby
LETTER_SCORES = {
  "A" => 1
  "B" => 3,
  "C" => 3,
  "D" => 2,
  # ...
}
```

Then to get the score for a letter, you can say `LETTER_SCORES[letter]`.

File: lib/adagrams.rb [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+score_list.each do |element|
+  if element[:score] > max
+    find_best_word = [element]
+    max = element[:score]
+  elsif element[:score] == max
+    find_best_word << element
+  end
+end
+p find_best_word
+winner = find_best_word.first
```
Comment (droberts-sea) [inline lib/adagrams.rb:101]: Please remove debugging output (like line 100) before submitting projects.

**Review by droberts-sea — COMMENTED** [review]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Small commits with meaningful commit messages | yes
**Code Requirements** |
`draw_letters` method |
Uses appropriate data structure to store the letter distribution | yes
All tests for `draw_letters` pass | yes
`uses_available_letters?` method |
All tests for `uses_available_letters?` pass | yes, but see inline
`score_word` method |
Uses appropriate data structure to store the letter scores | see inline
All tests for `score_word` pass | yes
`highest_score_from` method |
Appropriately handles edge cases for tie-breaking logic | yes
All tests for `highest_score_from` pass | yes
**Overall** |

Good job overall! I've left a few inline comments, and you need to get your indentation figured out, but in general I'm happy with what I see here. It is clear to me that the learning goals for this assignment were met. Keep up the hard work!
